### ISSUE TITLE
? operator for propagating Reslut, Option - does clippy give false positive warning ?

### ISSUE THREAD

**Issue by tomaszdrozdz** [rust-lang/rust-clippy#14675]
### Description

Hello all,

Is this warning OK - or false positive ?

```rust
let r: Result<i32, std::num::ParseIntError> = (|| -> Result<i32, std::num::ParseIntError> { 
    Ok("1".parse::<i32>()?) })();
```

Gives:
> question mark operator is useless here
for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_question_mark
`#[warn(clippy::needless_question_mark)]` on by default

----------------------
But of course:
```rust
let r: Result<i32, std::num::ParseIntError> = (|| -> Result<i32, std::num::ParseIntError> { 
    Ok("1".parse::<i32>() ) })();
```
Gives:
> mismatched types
expected type `i32`
   found enum `std::result::Result<i32, std::num::ParseIntError>`

----------------------
And this:
```rust
let r: Result<i32, std::num::ParseIntError> = (|| -> Result<i32, std::num::ParseIntError> { 
    Ok("1".parse::<i32>()? + "2".parse::<i32>()?) })();
```
does not produce warning.

### Version

```text

```

**Comment by Alexendoo** [comment]
Tweaked the message to be clearer in https://github.com/rust-lang/rust-clippy/pull/14682, the lint wants you to remove the `Ok` as well as the `?`

**Comment by tomaszdrozdz** [comment]
Thank You :)
